File: eb-application/src/main/java/com/ebnbin/ebapplication/net/NetModelCallback.java
package com.ebnbin.ebapplication.net;

import android.support.annotation.NonNull;

import com.ebnbin.ebapplication.model.EBModel;

import java.util.ArrayList;

import okhttp3.Call;

/**
 * Net callbacks for getting models.
 *
 * @param <Model> Subclass of {@link EBModel}.
 */
public abstract class NetModelCallback<Model extends EBModel> {
    /**
     * Adds extra pre-callbacks with this.
     */
    public final ArrayList<NetModelCallback<Model>> preCallbacks = new ArrayList<>();
    /**
     * Adds extra post-callbacks with this.
     */
    public final ArrayList<NetModelCallback<Model>> postCallbacks = new ArrayList<>();

    /**
     * Called on begin with pre-callbacks and post-callbacks.
     */
    final void onBeginCallback(@NonNull Call call) {
        for (NetModelCallback<Model> callback : preCallbacks) {
            callback.onBegin(call);
        }

        onBegin(call);

        for (NetModelCallback<Model> callback : postCallbacks) {
            callback.onBegin(call);
        }
    }

    /**
     * Called when success with pre-callbacks and post-callbacks.
     *
     * @param model Valid model.
     */
    final void onSuccessCallback(@NonNull Call call, @NonNull Model model) {
        for (NetModelCallback<Model> callback : preCallbacks) {
            callback.onSuccess(call, model);
        }

        onSuccess(call, model);

        for (NetModelCallback<Model> callback : postCallbacks) {
            callback.onSuccess(call, model);
        }
    }

    /**
     * Called when failure with pre-callbacks and post-callbacks.
     */
    final void onFailureCallback(@NonNull Call call) {
        for (NetModelCallback<Model> callback : preCallbacks) {
            callback.onFailure(call);
        }

        onFailure(call);

        for (NetModelCallback<Model> callback : postCallbacks) {
            callback.onFailure(call);
        }
    }

    /**
     * Called on end with pre-callbacks and post-callbacks.
     */
    final void onEndCallback(@NonNull Call call) {
        for (NetModelCallback<Model> callback : preCallbacks) {
            callback.onEnd(call);
        }

        onEnd(call);

        for (NetModelCallback<Model> callback : postCallbacks) {
            callback.onEnd(call);
        }
    }

    /**
     * Called on begin.
     */
    public void onBegin(@NonNull Call call) {
    }

    /**
     * Called when success.
     *
     * @param model Valid model.
     */
    public void onSuccess(@NonNull Call call, @NonNull Model model) {
    }

    /**
     * Called when failure.
     */
    public void onFailure(@NonNull Call call) {
    }

    /**
     * Called on end.
     */
    public void onEnd(@NonNull Call call) {
    }
}

File: eb-application/src/main/java/com/ebnbin/ebapplication/net/NetModelCallback.kt
package com.ebnbin.ebapplication.net

import com.ebnbin.ebapplication.model.EBModel
import okhttp3.Call
import okhttp3.Response
import java.io.IOException

/**
 * Net callback for getting a model.
 */
abstract class NetModelCallback<Model : EBModel> {
    /**
     * Called on begin.
     *
     * Count of [onBegin]
     * `=` count of [onSuccess] `+` count of [onFailure] `+` count of [onCancel]
     * `=` count of [onEnd].
     *
     * @see onSuccess
     * @see onFailure
     * @see onCancel
     * @see onEnd
     */
    open fun onBegin(call: Call) {}

    /**
     * Called on success.
     *
     * @param model Valid model.
     *
     * @see onBegin
     */
    open fun onSuccess(call: Call, model: Model, response: Response, byteArray: ByteArray) {}

    /**
     * Called on failure.
     *
     * @see onBegin
     */
    open fun onFailure(call: Call, errorCode: Int, e: IOException?, response: Response?) {}

    /**
     * Called on cancel.
     *
     * @see onBegin
     */
    open fun onCancel(call: Call) {}

    /**
     * Called on end.
     *
     * @see onBegin
     */
    open fun onEnd(call: Call) {}

    //*****************************************************************************************************************

    /**
     * Pre-callbacks.
     */
    val preCallbacks = mutableListOf<NetModelCallback<Model>>()
    /**
     * Post-callbacks.
     */
    val postCallbacks = mutableListOf<NetModelCallback<Model>>()

    /**
     * @see onBegin
     */
    internal fun begin(call: Call) {
        preCallbacks.forEach { preCallback -> preCallback.onBegin(call) }

        onBegin(call)

        postCallbacks.forEach { postCallback -> postCallback.onBegin(call) }
    }

    /**
     * @see onSuccess
     */
    internal fun success(call: Call, model: Model, response: Response, byteArray: ByteArray) {
        preCallbacks.forEach { preCallback -> preCallback.onSuccess(call, model, response, byteArray) }

        onSuccess(call, model, response, byteArray)

        postCallbacks.forEach { postCallback -> postCallback.onSuccess(call, model, response, byteArray) }
    }

    /**
     * @see onFailure
     */
    internal fun failure(call: Call, errorCode: Int, e: IOException?, response: Response?) {
        preCallbacks.forEach { preCallback -> preCallback.onFailure(call, errorCode, e, response) }

        onFailure(call, errorCode, e, response)

        postCallbacks.forEach { postCallback -> postCallback.onFailure(call, errorCode, e, response) }
    }

    /**
     * @see onCancel
     */
    internal fun cancel(call: Call) {
        preCallbacks.forEach { preCallback -> preCallback.onCancel(call) }

        onCancel(call)

        postCallbacks.forEach { postCallback -> postCallback.onCancel(call) }
    }

    /**
     * @see onEnd
     */
    internal fun end(call: Call) {
        preCallbacks.forEach { preCallback -> preCallback.onEnd(call) }

        onEnd(call)

        postCallbacks.forEach { postCallback -> postCallback.onEnd(call) }
    }

    //*****************************************************************************************************************

    companion object {
        //*************************************************************************************************************
        // Error codes.

        const val ERROR_CODE_WTF = 0
        const val ERROR_CODE_FAILURE = 1
        const val ERROR_CODE_RESPONSE_NULL = 2
        const val ERROR_CODE_RESPONSE_UNSUCCESSFUL = 3
        const val ERROR_CODE_RESPONSE_BODY_NULL = 4
        const val ERROR_CODE_JSON_SYNTAX_EXCEPTION = 5
        const val ERROR_CODE_CLASS_CAST_EXCEPTION = 6
        const val ERROR_CODE_MODEL_INVALID = 7
    }
}
